org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/MapStructCompletionProposalComputer.java

```diff
@@ -0,0 +1,279 @@
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.MemberValuePair;
+import org.eclipse.jdt.core.dom.MethodDeclaration;
+import org.eclipse.jdt.ui.ISharedImages;
+import org.eclipse.jdt.ui.JavaUI;
+import org.eclipse.jdt.ui.text.java.ContentAssistInvocationContext;
+import org.eclipse.jdt.ui.text.java.IJavaCompletionProposalComputer;
+import org.eclipse.jdt.ui.text.java.JavaContentAssistInvocationContext;
+import org.eclipse.jface.text.contentassist.CompletionProposal;
+import org.eclipse.jface.text.contentassist.ICompletionProposal;
+import org.eclipse.jface.text.contentassist.IContextInformation;
+
+/**
+ * Computes MapStruct specific content assist completion proposals.
+ * 
+ * @author Lars Wetzer
+ */
+public class MapStructCompletionProposalComputer implements IJavaCompletionProposalComputer {
+
+    static final String MAPPING_ANNOTATION_NAME = "Mapping"; //$NON-NLS-1$
+    static final String SOURCE_ANNOTATION_METHOD = "source"; //$NON-NLS-1$
+    static final String TARGET_ANNOTATION_METHOD = "target"; //$NON-NLS-1$
+    static final String GET_PREFIX = "get"; //$NON-NLS-1$
+    static final String SET_PREFIX = "set"; //$NON-NLS-1$
+
+    private static final List<IContextInformation> EMPTY_CONTEXTS = Collections.emptyList();
+    private static final List<ICompletionProposal> EMPTY_PROPOSALS = Collections.emptyList();
+
+    @Override
+    public void sessionStarted() {
+    }
+
+    @Override
+    public void sessionEnded() {
+    }
+
+    @Override
+    public String getErrorMessage() {
+        return null;
+    }
+
+    @Override
+    public List<IContextInformation> computeContextInformation(ContentAssistInvocationContext context,
+                                                               IProgressMonitor monitor) {
+        return EMPTY_CONTEXTS;
+    }
+
+    @Override
+    public List<ICompletionProposal> computeCompletionProposals(ContentAssistInvocationContext context,
+                                                                IProgressMonitor monitor) {
+
+        try {
+
+            if ( !( context instanceof JavaContentAssistInvocationContext ) ) {
+                return EMPTY_PROPOSALS;
+            }
+
+            JavaContentAssistInvocationContext javaContent = (JavaContentAssistInvocationContext) context;
+
+            ICompilationUnit compilationUnit = javaContent.getCompilationUnit();
+
+            if ( compilationUnit == null || !compilationUnit.isStructureKnown() ) {
+                return EMPTY_PROPOSALS;
+            }
+
+            int invocationOffset = javaContent.getInvocationOffset();
+
+            IJavaElement javaElement = compilationUnit.getElementAt( invocationOffset );
+
+            if ( !( javaElement instanceof IMethod ) ) {
+                return EMPTY_PROPOSALS;
+            }
+
+            IMethod method = (IMethod) javaElement;
+
+            for ( IAnnotation annotation : method.getAnnotations() ) {
```

Comment:
This seems not to cover the case where several mappings are given via `@Mappings`:

    @Mappings({
        @Mapping(source="foo", "target="bar"),
        @Mapping(source="")
                         ^--- No proposals here
    })
    Target toTarget( Source s );